contracts/Marketplace1155.sol

```diff
@@ -0,0 +1,161 @@
+
+    event MarketItemCreated(
+        uint256 indexed itemId,
+        address indexed nftContract,
+        uint256 indexed tokenIds,
+        address seller,
+        address owner,
+        uint256 price,
+        bool sold
+    );
+    // serve il bool sold? quando lo creo puo' essere gia' venduto?
+
+    event MarketItemSold(uint256 indexed itemId, address owner);
+
+    function onERC1155Received(
+        address,
+        address,
+        uint256,
+        uint256,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155Received.selector;
+    }
+
+    function onERC1155BatchReceived(
+        address,
+        address,
+        uint256[] memory,
+        uint256[] memory,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155BatchReceived.selector;
+    }
+
+    function createMarketItem(
+        address nftContract,
+        uint256[] memory tokenIds,
+        uint256 price,
+        uint256[] memory amounts // se e' un nft amounts non dovrebbe essere sempre 1? siccome immagino siano tracked dai tokenIds questo mi sembra superfluo
+    ) public payable nonReentrant {
+        require(price > 0, "Price must be greater than 0"); // Non vogliamo controllare che tokenIds sia lungo quanto amounts?
+
+        IERC1155(nftContract).safeBatchTransferFrom(
+            msg.sender,
+            address(this),
+            tokenIds,
+            amounts,
+            ""
+        );
+
+        for (uint256 i = 0; i <= (tokenIds.length - 1); i++) {
+            //davide: possiamo usare solo il < e togliere il -1?
+            _itemIds.increment();
+            uint256 itemId = _itemIds.current();
+
+            idToMarketItem[itemId] = MarketItem(
+                itemId,
+                nftContract,
+                tokenIds[i],
+                payable(msg.sender),
+                payable(address(0)),
+                price,
+                false
+            );
+
+            emit MarketItemCreated(
+                itemId,
+                nftContract,
+                tokenIds[i],
+                msg.sender,
+                address(0),
+                price,
+                false
+            );
+
+            // non vale la pena emettere un evento solo invece di n?
+        }
+    }
+
+    function createMarketSale(
+        address nftContract,
+        uint256 itemId,
+        uint256[] memory _tokenIds,
+        uint256[] memory amounts
+    ) public payable nonReentrant {
+        uint256 price = idToMarketItem[itemId].price;
+
+        bool sold = idToMarketItem[itemId].sold;
+        require(
+            msg.value == price,
+            "Please submit the asking price in order to complete the purchase"
+        );
+        require(sold != true, "This Sale has alredy finished"); //missing a
+        emit MarketItemSold(itemId, msg.sender); //Possiamo metterlo alla fine questo?
```

Comment:
We should move this to the end of function. If the function goes out of gas the event could still be emitted